### PR TITLE
eth/tracers: revert reason in call_tracer + error for failed internal calls

### PR DESCRIPTION
**Pull request by libotony**
Do the same thing as the struct logger does.
https://github.com/ethereum/go-ethereum/blob/a00dc5095b53c58136b0a9a1f19a7f0c98872cf4/eth/api_tracer.go#L767-L772

### CODE REVIEW

File: eth/tracers/internal/tracers/assets.go [hunk 1]
```diff
@@ -58,21 +55,32 @@ type bindataFileInfo struct {
...
 func (fi bindataFileInfo) Size() int64 {
 	return fi.size
 }
+
+// Mode return file mode
 func (fi bindataFileInfo) Mode() os.FileMode {
 	return fi.mode
 }
+
+// Mode return file modify time
```
Comment (MariusVanDerWijden) [inline eth/tracers/internal/tracers/assets.go:73]:
```suggestion
// ModeTime return file modify time
```

Comment (libotony) [reply eth/tracers/internal/tracers/assets.go:73]: updated my `go-bindata`, solved

File: eth/tracers/internal/tracers/assets.go [hunk 2]
```diff
@@ -46,9 +45,8 @@ func bindataRead(data []byte, name string) ([]byte, error) {
 }
 
 type asset struct {
-	bytes  []byte
-	info   os.FileInfo
-	digest [sha256.Size]byte
+	bytes []byte
+	info  os.FileInfo
```
Comment (holiman) [inline eth/tracers/internal/tracers/assets.go:49]: this doesn't seem to be using the same go-bindata as we've used previously. You don't have to fix that though, it's better if one of us regenerates it before merging.

File: eth/tracers/internal/tracers/call_tracer.js [hunk 3]
```diff
@@ -208,7 +208,7 @@
 		} else if (ctx.error !== undefined) {
 			result.error = ctx.error;
 		}
-		if (result.error !== undefined) {
+		if (result.error !== undefined && (result.error !== "execution reverted" || result.output ==="0x")) {
 			delete result.output;
 		}
```
Comment (holiman) [inline eth/tracers/internal/tracers/call_tracer.js:213]: I'm not sure why this `delete result.output` is needed at all, really. Any idea?

Comment (holiman) [reply eth/tracers/internal/tracers/call_tracer.js:213]: Deleting this statement should cause a test-failure on `tracers_test.go`, so we'll have to change `call_tracer_revert.json`, but that seems better to me

Comment (libotony) [reply eth/tracers/internal/tracers/call_tracer.js:213]: Regarding the `delete result.output`, I guess VM doesn't return any data before `revert(string memory message)` was added.

**Comment by karalabe** [comment]
We'll need a test for this. There are a few in `testdata` which were generated using the prestate_tracer against an existing chain. You could do a similar thing: pick any random transaction from any test network that reverted (with a reason), call a prestat_tracer on it and dump it into the `testdata`. You might need to combine it then with the expected result from a test run.

**Comment by libotony** [comment]
> We'll need a test for this. There are a few in `testdata` which were generated using the prestate_tracer against an existing chain. You could do a similar thing: pick any random transaction from any test network that reverted (with a reason), call a prestat_tracer on it and dump it into the `testdata`. You might need to combine it then with the expected result from a test run.

Thanks, I have added a new test file `eth/tracers/testdata/call_tracer_revert_reason.json`.

**Comment by holiman** [comment]
I changed the order of the commits, (to check the test agains the previous code), and pushed a commit on top to regenerate the assets. I have no idea why the evmdis-code apparently is not generated the same way it was previously, so I'll have to look into that.

**Comment by holiman** [comment]
Ah, evmdis tracer was modified here: https://github.com/ethereum/go-ethereum/commit/cc3ef1e4f48dc305c942bc2259b3121210206686#diff-988f7883cc7fdcb39f2e67e6681b4ec4L17 , but not regenerated (which is fine, since it was a no-op change). 
Interesting though -- that means we can shave off some bytes by stripping the javascript comments before packing it as an asset.

**Review by holiman — APPROVED** [review]

**Comment by holiman** [comment]
There are two more bugs I'd like to handle in this PR too, while we're doing work on the calltracer:

- https://github.com/ethereum/go-ethereum/issues/21489
- https://github.com/ethereum/go-ethereum/issues/21438

**Comment by holiman** [comment]
I'm 99% certain that both those issues are fixed with my latest commit, the root cause is the same. 

So this is how it would look now: 
```javascript
  "result": {
    "type": "CALL",
    "from": "0x66fdfd05e46126a07465ad24e40cc0597bc1ef31",
    "to": "0x6c06b16512b332e6cd8293a2974872674716ce18",
    "value": "0x0",
    "gas": "0x1a466",
    "gasUsed": "0x1dc6",
    "input": "0x2e1a7d4d00000000000000000000000000000000000000000000000014d1120d7b160000",
    "output": "0x",
    "calls": [
      {
        "type": "CALL",
        "from": "0x6c06b16512b332e6cd8293a2974872674716ce18",
        "to": "0x66fdfd05e46126a07465ad24e40cc0597bc1ef31",
        "value": "0x14d1120d7b160000",
        "error":"internal failure",
        "input": "0x"
      }
    ]
  }
```
The `value` is stlil present, meaning that an attempt to send `value` was made. However, there's also an `error` now, to show that the call failed. 

cc @mtbitcoin  @juliankoh is that an ok compromise?

**Comment by mtbitcoin** [comment]
@holiman That will work. I can test once its merged to master. Does this require a resynch of the node though?

**Comment by holiman** [comment]
> Does this require a resynch of the node though?

No. And actually, you _can_ try it out by sending the full js into the calltracer, so instead of using the bundled "calltracer" you provide the code yourself.

But otherwise I think we'll merge this pretty soon

**Comment by holiman** [comment]
If anyone is curious: 
```
[user@work go-ethereum]$ ls -la ./build/bin/geth-*js
-rwxrwxr-x 1 user user 38951760 Aug 27 10:42 ./build/bin/geth-compressjs
-rwxrwxr-x 1 user user 38959432 Aug 27 10:29 ./build/bin/geth-nocompressjs
```
It shaves off a whopping `7K` from `38Mb`, or `0.02%`  to minify the js. Not really worth pursuing

**Pull request merged**